### ISSUE TITLE
[Bug] Some `_path_to_file.py1.stats` files are generated in worksapace with anaconda

### ISSUE THREAD

**Issue by alkatar21** [microsoft/vscode-pylint#30]
For each file pylint generates messages for, such a file is created in the workspace. In a dev container I do not have this, but for several Conda environments where no pylint is installed and this extension is used.

**Comment by karthiknadig** [comment]
We don't pass any particular flags to generate stats. Looks like this is default on for pylint: https://pylint.pycqa.org/en/latest/technical_reference/features.html?highlight=persistent#general-options

We only pass these flags in the extension to pylint:
```
--reports=n --output-format=json
```

**Comment by alkatar21** [comment]
This is still the case in the current version (v2022.1.11051003). This does not appear to be fixed. I can't find it in the commits either.

**Comment by karthiknadig** [comment]
This is a pylint thing, we don't pass any flags to pylint asking to generate starts. It does this if you have persistence not disabled for performance purposes.

**Comment by alkatar21** [comment]
Why doesn't this happen in dev containers, but at least in all my conda envs if the default is to generate them?
I use the same global settings. I do not see where I have persistence not disabled?
What is the suggested way to bypass the generation with the extension? It is probably not the expected behavior of the extension to create such a stats file for every Python file and to clutter the worksace?

**Comment by karthiknadig** [comment]
It is not the expected behavior of the extension to do this. As I mentioned in my previous comment, we don't pass any flags in particular to enable this behavior. You can see exactly what we run in the logs, so we are not creating this file. or asking pylint to do so. I tested this locally with latest conda, I don't see any stat file being generated in the local project directory

As for why this might be happening. pylint usually tries to cache these results in a directory specified by `XDG_CACHE_HOME` or it picks default cache based on the OS. If for some reasons the permissions are not right or if the directory is not available, then it starts creating it locally in the same directory as the project. So, pylint is always creating this stat file just that normally it is not in the same directory as the project.

You can add `persistence=n` in your global pylint to disable it for all scenarios. For info on how pylint searches for config see here:
 https://pylint.pycqa.org/en/latest/user_guide/run.html#command-line-options

You can always add this to your args if nothing else is working:
```
`"pylint.args": ["--persistent=n"]
```

Here is more info on this:
https://stackoverflow.com/questions/66677007/file-stats-when-run-python-in-neovim
https://pylint.pycqa.org/en/latest/faq.html#where-is-the-persistent-data-stored-to-compare-between-successive-runs

**Comment by alkatar21** [comment]
Ok, I now understand the underlying problem. I thought the difference was the generating and not the location.
After a short analysis I found out that the folder `...\AppData\Local\pylint` is created when the extension is active, but then not used it seems. When I check a file with pylint in the VSCode console with `pylint --reports=n --output-format=json <file>.py`, the folder `...\AppData\Local\pylint` is used to store the stat file as expected. So there seems to be a difference between extension and console.

In the given stackoverflow [article](https://stackoverflow.com/a/68540775) the following is mentioned:
> The problem is, most likely, in the interaction of the calling plugin and the called pylint, because behavior with generating .stats files in the same directory with the checked files only works in this case.

So maybe it is a problem of the extension under Windows or certain conditions under Windows?

**Comment by alkatar21** [comment]
I have now also disabled the extension and tried the old way via the Python extension with `"python.linting.pylintEnabled": true,`. This also works as expected and the stats files are stored under `...\AppData\Local\pylint`. So it seems that this extension is the reason that the stats files are stored locally, especially since it works with the python extension.

**Comment by karthiknadig** [comment]
Can you provide the environment variables you get from an activated terminal (outside of VS code)?

also, install tis version of the extension: https://github.com/microsoft/vscode-pylint/suites/6164024713/artifacts/215951040
It should long the environment variables it sees. This way we can compare if there are any differences.

Additionally, can you test these two cases from VS Code terminal with the environment where this issue occurs:

### Case 1: How the python extension runs linter:

This is the Good case where it never generated the stat file.

1. create a file `test_linter.py`:
```
import subprocess
import sys

args = [
    sys.executable,
    "-m",
    "pylint",
    "--reports=n",
    "--output-format=json",
] + sys.argv[1:]
subprocess.run(args, encoding="utf-8", stdout=sys.stdout, stderr=sys.stderr)
```
2. run following command:
```
python.exe test_linter.py <some file with lint issues>
```

### Case 2: This is how the pylitn extension runs the linter:

This is the potential culprit case where the stat files are created.

1. create file `test_linter2.py`
```
import contextlib
import runpy
import sys


args = [
    "pylint",
    "--reports=n",
    "--output-format=json",
] + sys.argv[1:]


@contextlib.contextmanager
def switch_args(args):
    old = sys.argv[:]
    sys.argv = args[:]
    yield
    sys.argv[:] = old[:]


with switch_args(args):
    runpy.run_module("pylint", run_name="__main__")

```
2. run following command:
```
python.exe test_linter2.py <some file with lint issues>
```

If case 2 work well from an activated environment, try suing the python.exe from the environment but run it from an un-activated terminal. Like this:
```
.myvenv/Scripts/python.exe test_linter2.py <some file with lint issues>
```

This should give us clues into what is going on.

**Comment by alkatar21** [comment]
I am not quite sure what is meant by an activated terminal. I have now assumed one with Anaconda and active env:

I have adjusted the env slightly and removed other programs and replaced some paths with placeholders. This should be the same for all envs and not relevant anyway.

<details>
<summary>Conda Console env</summary>

```
(images) PS D:\> Get-ChildItem Env:

Name                           Value
----                           -----
_CONDA_EXE                     D:/Programme/Anaconda3\Scripts\conda.exe
_CONDA_ROOT                    D:/Programme/Anaconda3
ALLUSERSPROFILE                C:\ProgramData
APPDATA                        C:\Users\<username>\AppData\Roaming
CommonProgramFiles             C:\Program Files\Common Files
CommonProgramFiles(x86)        C:\Program Files (x86)\Common Files
CommonProgramW6432             C:\Program Files\Common Files
ComSpec                        C:\WINDOWS\system32\cmd.exe
CONDA_DEFAULT_ENV              images
CONDA_EXE                      D:\Programme\Anaconda3\Scripts\conda.exe
CONDA_PREFIX                   D:\Programme\Anaconda3\envs\images
CONDA_PREFIX_1                 D:\Programme\Anaconda3
CONDA_PROMPT_MODIFIER          (images)
CONDA_PYTHON_EXE               D:\Programme\Anaconda3\python.exe
CONDA_SHLVL                    2
DriverData                     C:\Windows\System32\Drivers\DriverData
HOMEDRIVE                      C:
HOMEPATH                       \Users\<username>
LOCALAPPDATA                   C:\Users\<username>\AppData\Local
NUMBER_OF_PROCESSORS           8
OneDrive                       C:\Users\<username>\OneDrive
OS                             Windows_NT
Path                           D:\Programme\Anaconda3\envs\images;D:\Programme\Anaconda3\envs\images\Library\mingw-w...
PATHEXT                        .COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL
ProgramData                    C:\ProgramData
ProgramFiles                   C:\Program Files
ProgramFiles(x86)              C:\Program Files (x86)
ProgramW6432                   C:\Program Files
PSExecutionPolicyPreference    Bypass
PSModulePath                   C:\Users\<username>\Documents\WindowsPowerShell\Modules;C:\Program Files\WindowsPowerShe...
PUBLIC                         C:\Users\Public
SystemDrive                    C:
SystemRoot                     C:\WINDOWS
TEMP                           C:\Users\<username>\AppData\Local\Temp
TMP                            C:\Users\<username>\AppData\Local\Temp
USERNAME                       <username>
USERPROFILE                    C:\Users\<username>
windir                         C:\WINDOWS
WSLENV                         WT_SESSION::WT_PROFILE_ID
WT_PROFILE_ID                  {c18b55cf-2420-4642-a273-63311ec98702}
WT_SESSION                     08c3ec15-c44a-49db-93ed-ff203920bd61

(images) PS D:\> $Env:PATH
D:\Programme\Anaconda3\envs\images;D:\Programme\Anaconda3\envs\images\Library\mingw-w64\bin;D:\Programme\Anaconda3\envs\images\Library\usr\bin;D:\Programme\Anaconda3\envs\images\Library\bin;D:\Programme\Anaconda3\envs\images\Scripts;D:\Programme\Anaconda3\envs\images\bin;D:\Programme\Anaconda3\condabin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0;C:\WINDOWS\System32\OpenSSH;C:\Program Files\dotnet;D:\Programme\Anaconda3;D:\Programme\Anaconda3\Library\mingw-w64\bin;D:\Programme\Anaconda3\Library\usr\bin;D:\Programme\Anaconda3\Library\bin;D:\Programme\Anaconda3\Scripts;C:\Users\<username>\AppData\Local\Microsoft\WindowsApps;D:\Programme\VSCode\bin
```
</details>

I installed the extension and thats the env output:

<details>
<summary>Extension env Output</summary>

```
sys.path used to run Linter:
   c:\Users\<username>\.vscode\extensions\ms-python.pylint-2022.1.2185866462-dev\bundled\linter
   D:\Programme\Anaconda3\envs\images\python310.zip
   D:\Programme\Anaconda3\envs\images\DLLs
   D:\Programme\Anaconda3\envs\images\lib
   D:\Programme\Anaconda3\envs\images
   D:\Programme\Anaconda3\envs\images\lib\site-packages
   c:\Users\<username>\.vscode\extensions\ms-python.pylint-2022.1.2185866462-dev\bundled\libs
Environment variables used:
    ALLUSERSPROFILE=C:\ProgramData
   APPDATA=C:\Users\<username>\AppData\Roaming
   APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL=true
   CHROME_CRASHPAD_PIPE_NAME=\\.\pipe\crashpad_17620_GIWKHOPZYOUOTXII
   COMMONPROGRAMFILES=C:\Program Files\Common Files
   COMMONPROGRAMFILES(X86)=C:\Program Files (x86)\Common Files
   COMMONPROGRAMW6432=C:\Program Files\Common Files
   COMSPEC=C:\WINDOWS\system32\cmd.exe
   DRIVERDATA=C:\Windows\System32\Drivers\DriverData
   ELECTRON_NO_ATTACH_CONSOLE=1
   ELECTRON_RUN_AS_NODE=1
   HOMEDRIVE=C:
   HOMEPATH=\Users\<username>
   LOCALAPPDATA=C:\Users\<username>\AppData\Local
   LOGONSERVER=\\<username>-LAPTOP
   NUMBER_OF_PROCESSORS=8
   ONEDRIVE=C:\Users\<username>\OneDrive
   ORIGINAL_XDG_CURRENT_DESKTOP=undefined
   OS=Windows_NT
   PATH=C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\dotnet\;D:\Programme\Anaconda3;D:\Programme\Anaconda3\Library\mingw-w64\bin;D:\Programme\Anaconda3\Library\usr\bin;D:\Programme\Anaconda3\Library\bin;D:\Programme\Anaconda3\Scripts;C:\Users\<username>\AppData\Local\Microsoft\WindowsApps;D:\Programme\VSCode\bin;
   PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
   PROGRAMDATA=C:\ProgramData
   PROGRAMFILES=C:\Program Files
   PROGRAMFILES(X86)=C:\Program Files (x86)
   PROGRAMW6432=C:\Program Files
   PROMPT=$P$G
   PSMODULEPATH=C:\Program Files\WindowsPowerShell\Modules;C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules
   PUBLIC=C:\Users\Public
   SESSIONNAME=Console
   SYSTEMDRIVE=C:
   SYSTEMROOT=C:\WINDOWS
   TEMP=C:\Users\<username>\AppData\Local\Temp
   TMP=C:\Users\<username>\AppData\Local\Temp
   USERNAME=<username>
   USERPROFILE=C:\Users\<username>
   VSCODE_AMD_ENTRYPOINT=vs/workbench/api/node/extensionHostProcess
   VSCODE_CLI=1
   VSCODE_CODE_CACHE_PATH=C:\Users\<username>\AppData\Roaming\Code\CachedData\dfd34e8260c270da74b5c2d86d61aee4b6d56977
   VSCODE_CWD=<path/to/project>
   VSCODE_HANDLES_UNCAUGHT_ERRORS=true
   VSCODE_IPC_HOOK=\\.\pipe\321d4d203506c48b68e7ce317f28ca55-1.66.2-main-sock
   VSCODE_IPC_HOOK_EXTHOST=\\.\pipe\vscode-ipc-09a7a3a0-6dd5-42f7-ab6a-38c7ae634a13-sock
   VSCODE_LOG_NATIVE=false
   VSCODE_LOG_STACK=false
   VSCODE_NLS_CONFIG={"locale":"en-us","availableLanguages":{},"_languagePackSupport":true}
   VSCODE_PID=17620
   VSCODE_PIPE_LOGGING=true
   VSCODE_VERBOSE_LOGGING=true
   WINDIR=C:\WINDOWS
```
</details>

I also used the following code to generate the output of the env on the VSCode console:
`get_env.py`:
```py
import os
envs = "\r\n   ".join(f"{k}={os.environ[k]}" for k in os.environ)
print(f"Environment variables used:\r\n    {envs}")
```
`python.exe ./get_env.py`

My console is defined as follows:
```jsonc
    "terminal.integrated.profiles.windows": {
        "Anaconda": {
            "overrideName": true,
            "source": "PowerShell",
            "args": [
                "-ExecutionPolicy",
                "ByPass",
                "-NoExit",
                "-Command",
                "D:/Programme/Anaconda3/shell/condabin/conda-hook.ps1",
                ";conda activate 'D:/Programme/Anaconda3'"
            ],
        },
    },
    "python.terminal.activateEnvInCurrentTerminal": true,
```

<details>
<summary>VSCode Console env</summary>

```
Environment variables used:
    ALLUSERSPROFILE=C:\ProgramData
   APPDATA=C:\Users\<username>\AppData\Roaming
   APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL=true
   CHROME_CRASHPAD_PIPE_NAME=\\.\pipe\crashpad_17620_GIWKHOPZYOUOTXII
   COMMONPROGRAMFILES=C:\Program Files\Common Files
   COMMONPROGRAMFILES(X86)=C:\Program Files (x86)\Common Files
   COMMONPROGRAMW6432=C:\Program Files\Common Files
   COMSPEC=C:\WINDOWS\system32\cmd.exe
   CONDA_DEFAULT_ENV=images
   CONDA_EXE=D:\Programme\Anaconda3\Scripts\conda.exe
   CONDA_PREFIX=D:\Programme\Anaconda3\envs\images
   CONDA_PREFIX_1=D:\Programme\Anaconda3
   CONDA_PROMPT_MODIFIER=(images)
   CONDA_PYTHON_EXE=D:\Programme\Anaconda3\python.exe
   CONDA_SHLVL=2
   DRIVERDATA=C:\Windows\System32\Drivers\DriverData
   HOMEDRIVE=C:
   HOMEPATH=\Users\<username>
   LOCALAPPDATA=C:\Users\<username>\AppData\Local
   NUMBER_OF_PROCESSORS=8
   ONEDRIVE=C:\Users\<username>\OneDrive
   ORIGINAL_XDG_CURRENT_DESKTOP=undefined
   OS=Windows_NT
   PATH=D:\Programme\Anaconda3\envs\images;D:\Programme\Anaconda3\envs\images\Library\mingw-w64\bin;D:\Programme\Anaconda3\envs\images\Library\usr\bin;D:\Programme\Anaconda3\envs\images\Library\bin;D:\Programme\Anaconda3\envs\images\Scripts;D:\Programme\Anaconda3\envs\images\bin;D:\Programme\Anaconda3\condabin;C:\Program Files\WindowsApps\Microsoft.PowerShell_7.2.2.0_x64__8wekyb3d8bbwe;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0;C:\WINDOWS\System32\OpenSSH;C:\Program Files\dotnet;D:\Programme\Anaconda3;D:\Programme\Anaconda3\Library\mingw-w64\bin;D:\Programme\Anaconda3\Library\usr\bin;D:\Programme\Anaconda3\Library\bin;D:\Programme\Anaconda3\Scripts;C:\Users\<username>\AppData\Local\Microsoft\WindowsApps;D:\Programme\VSCode\bin;.
   PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL
   PROGRAMDATA=C:\ProgramData
   PROGRAMFILES=C:\Program Files
   PROGRAMFILES(X86)=C:\Program Files (x86)
   PROGRAMW6432=C:\Program Files
   PROMPT=$P$G
   PSEXECUTIONPOLICYPREFERENCE=Bypass
   PSMODULEPATH=C:\Users\<username>\Documents\PowerShell\Modules;C:\Program Files\PowerShell\Modules;c:\program files\windowsapps\microsoft.powershell_7.2.2.0_x64__8wekyb3d8bbwe\Modules;C:\Program Files\WindowsPowerShell\Modules;C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules
   PUBLIC=C:\Users\Public
   SESSIONNAME=Console
   SYSTEMDRIVE=C:
   SYSTEMROOT=C:\WINDOWS
   TEMP=C:\Users\<username>\AppData\Local\Temp
   TMP=C:\Users\<username>\AppData\Local\Temp
   USERNAME=<username>
   USERPROFILE=C:\Users\<username>
   WINDIR=C:\WINDOWS
   TERM_PROGRAM=vscode
   TERM_PROGRAM_VERSION=1.66.2
   LANG=en_US.UTF-8
   COLORTERM=truecolor
   _CONDA_EXE=D:/Programme/Anaconda3\Scripts\conda.exe
   _CONDA_ROOT=D:/Programme/Anaconda3
```
</details>

I also tested all the specified cases. For the enabled ones, I used the already specified Conda Console in VSCode. For the non-enabled attempt I used the default Powerschell profile in VSCode with 
```
D:\Programme\Anaconda3\envs\images\python.exe .\test_linter2.py <some file with lint issues>
```

In all cases the stat file was correctly written to the `...\AppData\Local\pylint` folder.

Since the `...\AppData\Local\pylint` folder is created with the extension, but then not used for the stats files when I save and thus trigger the linting is there perhaps still a difference somewhere?

**Comment by alkatar21** [comment]
I unzipped the extension and added [here](https://github.com/microsoft/vscode-pylint/blob/725fc7ae415a9638e2a44e922cf5133efde62811/bundled/linter/linter_server.py#L157) the following:
```py
LSP_SERVER.show_message_log(f"Use Path: {use_path}, {settings}")
```
`Use Path: False, {'workspace': 'file:///d%3A/<path>', 'trace': 'error', 'args': ['--enable-all-extensions', '--enable=useless-suppression'], 'severity': {'convention': 'Information', 'error': 'Error', 'fatal': 'Error', 'refactor': 'Information', 'warning': 'Warning', 'info': 'Information'}, 'path': [], 'interpreter': ['D:/Programme/Anaconda3/envs/images/python.exe'], 'workspaceFS': 'd:\\<path>'}`

And [here](https://github.com/microsoft/vscode-pylint/blob/725fc7ae415a9638e2a44e922cf5133efde62811/bundled/linter/linter_server.py#L168):
```py
        with utils.substitute_attr(sys, "path", sys.path[:]):
            LSP_SERVER.show_message_log(f"Path: {sys.path}")
```

`Path: ['c:\\Users\\<username>\\.vscode\\extensions\\ms-python.pylint-2022.1.2185866462-dev\\bundled\\linter', 'D:\\Programme\\Anaconda3\\envs\\images\\python310.zip', 'D:\\Programme\\Anaconda3\\envs\\images\\DLLs', 'D:\\Programme\\Anaconda3\\envs\\images\\lib', 'D:\\Programme\\Anaconda3\\envs\\images', 'D:\\Programme\\Anaconda3\\envs\\images\\lib\\site-packages', 'c:\\Users\\<username>\\.vscode\\extensions\\ms-python.pylint-2022.1.2185866462-dev\\bundled\\libs']`

Maybe that helps in some way?

**Comment by karthiknadig** [comment]
@alkatar21 Thanks for trying these out. I suspect the problem is not really with the environment variables or permissions, I believe it is with how we invoke the linter. We use stdin to works with source files, this is to allow scenarios where we may not have saved to disk yet. Somehow that might be causing pylint to use the current working directory.


One more thing to test

1. create file `test_linter2.py`
```
import contextlib
import io
import pathlib
import runpy
import sys
from typing import Any, Sequence


file_path = pathlib.Path(sys.argv[1])
args = ["pylint", "--reports=n", "--output-format=json", "--from-stdin", str(file_path)]


class CustomIO(io.TextIOWrapper):
    name = None

    def __init__(self, name, encoding="utf-8", newline=None):
        self._buffer = io.BytesIO()
        self._buffer.name = name
        super().__init__(self._buffer, encoding=encoding, newline=newline)

    def close(self):
        # This is intentionally empty.
        pass

    def get_value(self) -> str:
        self.seek(0)
        return self.read()


@contextlib.contextmanager
def substitute_attr(obj: Any, attribute: str, new_value: Any):
    old_value = getattr(obj, attribute)
    setattr(obj, attribute, new_value)
    yield
    setattr(obj, attribute, old_value)


@contextlib.contextmanager
def redirect_io(stream: str, new_stream):
    old_stream = getattr(sys, stream)
    setattr(sys, stream, new_stream)
    yield
    setattr(sys, stream, old_stream)


def run_module(module: str, argv: Sequence[str], source: str = None):
    try:
        with substitute_attr(sys, "argv", argv):
            str_input = CustomIO("<stdin>", encoding="utf-8", newline="\n")
            with redirect_io("stdin", str_input):
                str_input.write(source)
                str_input.seek(0)
                runpy.run_module(module, run_name="__main__")
    except SystemExit:
        pass


run_module("pylint", args, file_path.read_text())

```
2. run following command:
```
python.exe test_linter2.py <some file with lint issues>
```

**Comment by karthiknadig** [comment]
@alkatar21 I will need some input from pylint to debug this further. I have asked the question there.

**Comment by karthiknadig** [comment]
@alkatar21 Thanks for the patience and for trying things out. Since I can't repro this locally, I have to rely on you for this. Can you try this and see what it prints out?
https://github.com/PyCQA/pylint/issues/6394#issuecomment-1103093792

This will probably cause parse failures, so the extension might crash but it will print this out in the logs. You have to edit the pylint file that is in the extensions folder (if you are not using your pylint from your environment). This should be available under `%USEPROFILE%\.vscode\extensions\ms-python.pylint-2022.1.10811002\bundled\libs\pylint\config\__init__.py`

**Comment by alkatar21** [comment]
I would have tried more, but I didn't know where print statements would end up and I don't really have time to get into `pylint` to find the relevant functions. If I know where and how to test something then I'm happy to do it.  The time zones are not optimal as I am in Europe, but that just makes it take a bit longer.

Output while using the pylint extension:
```
base='d:\\Test\\test_linter.py'
PYLINT_HOME='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache'
data_file='d:_Test_test_linter.py1.stats'
```

Output using my VSCode console:
```
base='test_linter.py'
PYLINT_HOME='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache'
data_file='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache\\test_linter.py1.stats'
```

In addition, the following is in the DevContainer where the extension works as expected:
The extension:
```
base='/workspace/<path-to-file>/test.py'
PYLINT_HOME='/root/.cache/pylint'
data_file='/root/.cache/pylint/_workspace_path_to_file_test.py1.stats'
```
Unix console with `pylint test.py`:
```
base='test'
PYLINT_HOME='/root/.cache/pylint'
data_file='/root/.cache/pylint/test1.stats'
```
Here both are working as expected.

### Results

So I think we found the bug. There are 2 differences, one os specific and one between testing(console) and extension.
The OS specific line is the following:
https://github.com/PyCQA/pylint/blob/ae5ed5c57a8f9a60a37bd010be052b08aa864de7/pylint/config/__init__.py#L82
```py
# For the extension path on windows:
>>> os.path.join('C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache', 'd:_Test_test_linter.py1.stats')                  
'd:_Test_test_linter.py1.stats'
# For the extension path on unix:
>>> os.path.join('/root/.cache/pylint', '_workspace_path_to_file_test.py1.stats')
'/root/.cache/pylint/_workspace_path_to_file_test.py1.stats.py'
```
The drive letter swallows `PYLINT_HOME` under windows.

The other difference is that I used relative paths for testing, but the extension used absolute paths.
With that in mind, I ran the tests again:

```bash
(images) PS D:\Test> python.exe .\test_linter.py D:\Test\test_linter.py                                   
base='test_linter'
PYLINT_HOME='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache'
data_file='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache\\test_linter1.stats'

(images) PS D:\Test> python.exe .\test_linter2.py D:\Test\test_linter.py
base='test_linter'
PYLINT_HOME='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache'
data_file='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache\\test_linter1.stats'

# I called your second (longer) version of test_linter2.py test_linter3.py
(images) PS D:\Test> python.exe .\test_linter3.py D:\Test\test_linter.py
base='D:\\Test\\test_linter.py'
PYLINT_HOME='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache'
data_file='D:_Test_test_linter.py1.stats'
```

I think this information will help.

**Comment by DanielNoord** [comment]
> Output while using the pylint extension:
> 
> ```
> base='d:\\Test\\test_linter.py'
> PYLINT_HOME='C:\\Users\\<username>\\AppData\\Local\\pylint\\pylint\\Cache'
> data_file='d:_Test_test_linter.py1.stats'
> ```
> 

This seems to confirm what I thought, here https://github.com/PyCQA/pylint/issues/6394#issuecomment-1103093792.

I guess we don't handle drive letters in `FileItem.base_name` correctly. I will probably have time to look at this somewhere this week and we should be able to fix this in `2.14`.
If somebody else wants to take a look at this, please go ahead. I have quite a bit of other issues to fix before we can release `2.14`.

_The fix could obviously also be to remove the drive letter before creating a `FileItem` but I need to look at our internals to see when and how we decide `base_name` as I don't know that by heart._

**Comment by DanielNoord** [comment]
Sorry for taking a little longer to respond to this.

@alkatar21 Are you still interested in this issue? I don't have easy access to a similar test environment so I would like to ask you if you could test a fix I came up with. 
Could you test if this works:

```diff
diff --git a/pylint/config/__init__.py b/pylint/config/__init__.py
index cc82010b7..ef3b7e064 100644
--- a/pylint/config/__init__.py
+++ b/pylint/config/__init__.py
@@ -80,7 +80,7 @@ else:
 
 
 def _get_pdata_path(base_name: str, recurs: int) -> pathlib.Path:
-    base_name = base_name.replace(os.sep, "_")
+    base_name = os.path.splitdrive(base_name.replace(os.sep, "_"))[1]
     return pathlib.Path(PYLINT_HOME) / f"{base_name}{recurs}.stats"
 
```

**Comment by DanielNoord** [comment]
That's good to hear! 

Looking at the output I am going to change the fix a little bit. I think we should still keep the `D` somewhere in the `data_file` path as we don't want to confuse files in similar folders on different drives.

I'm not sure how proficient you are with `pytest` and the like, but any help getting a regression test for this for `pylint` would be much appreciated. I'll ping you when I submit the PR. I need to find a way to simulate different drive mounts in `pytest`...

**Comment by DanielNoord** [comment]
@alkatar21 We actually completely refactored the relevant code today so the previous `diff` no longer works.

I have created https://github.com/DanielNoord/pylint/pull/135 and used your suggestion for a test. Still, I'd like a real world test.
Could you test if that change works? You can clone the branch of the PR and `cd` into the directory and use `pip install -e .` to install that version into your environment (assuming that you have activated your environment before).

**Comment by DanielNoord** [comment]
I have just merged https://github.com/PyCQA/pylint/pull/6526.

Due to the recent refactor and the fact that `2.14` is nearing its release we decided not to back port this to a patch release. Since we have never seen issues about this we don't think it is a very widespread issue. The PR and this issue do have `diffs` that solve the issue on `2.13.x` and below.

I'll leave it up to the maintainers here whether this issue should stay open for visibility until `2.14` is released.

Thanks all for helping narrowing this down. Glad we found the issue eventually 🎉

**Comment by alkatar21** [comment]
I think the main reason why this has not been noticed so far is because no one else calls `pylint` the way this extension does it now. The python extension does it also differently. Since there are only 16k downloads so far, it has not been noticed, especially since it is not known who uses it at all and uses windows with 2 hard drives.
But as soon as the extension becomes much more popular, there would probably have been many more cases.

**Comment by DanielNoord** [comment]
Just checking in: we have released a beta for `2.14` and will likely release the actual version this week.

**Comment by karthiknadig** [comment]
The extension is now updated to use pylint 2.14.
